File: packages/schema/test/pointer.test.ts
import { describe, expect, it } from "vitest";
import {
  BASELINE_CONFIG,
  formatPointer,
  getAtPath,
  parsePointer,
  removeAtPath,
  setAtPath,
} from "../src/index.js";

const at = (pointer: string): readonly string[] => {
  const parsed = parsePointer(pointer);
  if (!parsed.ok) throw new Error(parsed.error.message);
  return parsed.value;
};

describe("pointer", () => {
  it("round-trips pointers with escapes", () => {
    expect(formatPointer(at("/a~1b/c~0d"))).toBe("/a~1b/c~0d");
    expect(at("/a~1b/c~0d")).toEqual(["a/b", "c~d"]);
  });

  it("rejects pointers that do not start with a slash", () => {
    expect(parsePointer("spec/launcher").ok).toBe(false);
  });

  it("reads scalars", () => {
    const result = getAtPath(BASELINE_CONFIG, at("/spec/launcher/columns"));
    expect(result.ok && result.value).toBe(1);
  });

  it("addresses keyed collections by key rather than index", () => {
    const result = getAtPath(BASELINE_CONFIG, at("/spec/apps/entries/com.spotify.music/network"));
    expect(result.ok && result.value).toBe("allow");
  });

  it("still accepts numeric indices", () => {
    const result = getAtPath(BASELINE_CONFIG, at("/spec/apps/entries/0/package"));
    expect(result.ok && result.value).toBe("os.normal.phone");
  });

  it("reports a missing key", () => {
    const result = getAtPath(BASELINE_CONFIG, at("/spec/apps/entries/com.example.ghost"));
    expect(result.ok).toBe(false);
    if (result.ok) return;
    expect(result.error.code).toBe("not-found");
  });

  it("sets without mutating the source", () => {
    const before = JSON.stringify(BASELINE_CONFIG);
    const updated = setAtPath(BASELINE_CONFIG, at("/spec/launcher/columns"), 3);
    expect(updated.ok).toBe(true);
    expect(JSON.stringify(BASELINE_CONFIG)).toBe(before);
    if (!updated.ok) return;
    const read = getAtPath(updated.value, at("/spec/launcher/columns"));
    expect(read.ok && read.value).toBe(3);
  });

  it("appends to a keyed collection when the key is new", () => {
    const entry = {
      package: "org.fdroid.fdroid",
      source: "fdroid",
      state: "installed",
      network: "wifi-only",
      permissions: {},
    };
    const updated = setAtPath(BASELINE_CONFIG, at("/spec/apps/entries/org.fdroid.fdroid"), entry);
    expect(updated.ok).toBe(true);
    if (!updated.ok) return;
    const read = getAtPath(updated.value, at("/spec/apps/entries/org.fdroid.fdroid/network"));
    expect(read.ok && read.value).toBe("wifi-only");
  });

  it("removes a keyed element", () => {
    const updated = removeAtPath(BASELINE_CONFIG, at("/spec/apps/entries/com.spotify.music"));
    expect(updated.ok).toBe(true);
    if (!updated.ok) return;
    expect(getAtPath(updated.value, at("/spec/apps/entries/com.spotify.music")).ok).toBe(false);
    expect(BASELINE_CONFIG.spec.apps.entries.length).toBe(6);
  });

  it("refuses to traverse into a scalar", () => {
    const result = setAtPath(BASELINE_CONFIG, at("/spec/launcher/columns/nope"), 1);
    expect(result.ok).toBe(false);
    if (result.ok) return;
    expect(result.error.code).toBe("not-traversable");
  });
});

File: pkg/config/pointer_test.go
package config_test

import (
	"reflect"
	"testing"

	"github.com/brandonapol/normal/pkg/config"
)

func at(t *testing.T, pointer string) []string {
	t.Helper()
	segments, err := config.ParsePointer(pointer)
	if err != nil {
		t.Fatalf("ParsePointer(%q): %v", pointer, err)
	}
	return segments
}

func document(t *testing.T) map[string]any {
	t.Helper()
	doc, err := config.ToDocument(config.Baseline())
	if err != nil {
		t.Fatalf("ToDocument: %v", err)
	}
	return doc
}

func TestPointerRoundTrip(t *testing.T) {
	segments := at(t, "/a~1b/c~0d")
	if !reflect.DeepEqual(segments, []string{"a/b", "c~d"}) {
		t.Fatalf("unexpected segments %v", segments)
	}
	if got := config.FormatPointer(segments); got != "/a~1b/c~0d" {
		t.Fatalf("round trip produced %q", got)
	}
}

func TestPointerMustStartWithSlash(t *testing.T) {
	if _, err := config.ParsePointer("spec/launcher"); err == nil {
		t.Fatal("expected an error")
	}
}

func TestGetScalar(t *testing.T) {
	value, err := config.GetAtPath(document(t), at(t, "/spec/launcher/columns"))
	if err != nil {
		t.Fatalf("GetAtPath: %v", err)
	}
	if value != float64(1) {
		t.Fatalf("expected 1, got %v (%T)", value, value)
	}
}

func TestGetByCollectionKey(t *testing.T) {
	value, err := config.GetAtPath(document(t), at(t, "/spec/apps/entries/com.spotify.music/network"))
	if err != nil {
		t.Fatalf("GetAtPath: %v", err)
	}
	if value != "allow" {
		t.Fatalf("expected allow, got %v", value)
	}
}

func TestGetByNumericIndexStillWorks(t *testing.T) {
	value, err := config.GetAtPath(document(t), at(t, "/spec/apps/entries/0/package"))
	if err != nil {
		t.Fatalf("GetAtPath: %v", err)
	}
	if value != "os.normal.phone" {
		t.Fatalf("expected the first entry, got %v", value)
	}
}

func TestGetMissingKeyIsNotFound(t *testing.T) {
	_, err := config.GetAtPath(document(t), at(t, "/spec/apps/entries/com.example.ghost"))
	pointerErr, ok := err.(*config.PointerError)
	if !ok || pointerErr.Code != config.ErrNotFound {
		t.Fatalf("expected not-found, got %v", err)
	}
}

func TestSetDoesNotMutateTheSource(t *testing.T) {
	source := document(t)
	updated, err := config.SetAtPath(source, at(t, "/spec/launcher/columns"), 3)
	if err != nil {
		t.Fatalf("SetAtPath: %v", err)
	}
	original, _ := config.GetAtPath(source, at(t, "/spec/launcher/columns"))
	if original != float64(1) {
		t.Fatalf("source was mutated, columns is now %v", original)
	}
	changed, _ := config.GetAtPath(updated, at(t, "/spec/launcher/columns"))
	if changed != 3 {
		t.Fatalf("expected 3, got %v", changed)
	}
}

func TestSetAppendsToKeyedCollection(t *testing.T) {
	entry := map[string]any{
		"package":     "org.fdroid.fdroid",
		"source":      "fdroid",
		"state":       "installed",
		"network":     "wifi-only",
		"permissions": map[string]any{},
	}
	updated, err := config.SetAtPath(document(t), at(t, "/spec/apps/entries/org.fdroid.fdroid"), entry)
	if err != nil {
		t.Fatalf("SetAtPath: %v", err)
	}
	value, err := config.GetAtPath(updated, at(t, "/spec/apps/entries/org.fdroid.fdroid/network"))
	if err != nil || value != "wifi-only" {
		t.Fatalf("expected the appended entry to be addressable, got %v (%v)", value, err)
	}
}

func TestRemoveKeyedElement(t *testing.T) {
	source := document(t)
	updated, err := config.RemoveAtPath(source, at(t, "/spec/apps/entries/com.spotify.music"))
	if err != nil {
		t.Fatalf("RemoveAtPath: %v", err)
	}
	if _, err := config.GetAtPath(updated, at(t, "/spec/apps/entries/com.spotify.music")); err == nil {
		t.Fatal("entry should be gone")
	}
	if _, err := config.GetAtPath(source, at(t, "/spec/apps/entries/com.spotify.music")); err != nil {
		t.Fatal("source document was mutated")
	}
}

func TestCannotTraverseIntoScalar(t *testing.T) {
	_, err := config.SetAtPath(document(t), at(t, "/spec/launcher/columns/nope"), 1)
	pointerErr, ok := err.(*config.PointerError)
	if !ok || pointerErr.Code != config.ErrNotTraversable {
		t.Fatalf("expected not-traversable, got %v", err)
	}
}
